pkg/containerutil/cp_linux.go

```diff
@@ -237,97 +277,74 @@ func CopyFiles(ctx context.Context, client *containerd.Client, container contain
 		return err
 	}
 	tarXCmd.Stdout = os.Stderr
-	tarXCmd.Stderr = os.Stderr
+	var tarErr bytes.Buffer
+	tarXCmd.Stderr = &tarErr
 
 	log.G(ctx).Debugf("executing %v in %q", tarCCmd.Args, tarCCmd.Dir)
 	if err := tarCCmd.Start(); err != nil {
-		return fmt.Errorf("failed to execute %v: %w", tarCCmd.Args, err)
+		return errors.Join(fmt.Errorf("failed to execute %v", tarCCmd.Args), err)
 	}
+
 	log.G(ctx).Debugf("executing %v in %q", tarXCmd.Args, tarXCmd.Dir)
 	if err := tarXCmd.Start(); err != nil {
-		return fmt.Errorf("failed to execute %v: %w", tarXCmd.Args, err)
+		if strings.Contains(err.Error(), "permission denied") {
+			return ErrTargetIsReadOnly
+		}
+
+		// Other errors, just put them back on stderr
+		_, fpErr := fmt.Fprint(os.Stderr, tarErr.String())
+		if fpErr != nil {
+			return errors.Join(fpErr, err)
+		}
+
+		return errors.Join(fmt.Errorf("failed to execute %v", tarXCmd.Args), err)
 	}
+
 	if err := tarCCmd.Wait(); err != nil {
 		return fmt.Errorf("failed to wait %v: %w", tarCCmd.Args, err)
 	}
+
 	if err := tarXCmd.Wait(); err != nil {
-		return fmt.Errorf("failed to wait %v: %w", tarXCmd.Args, err)
+		if strings.Contains(tarErr.String(), "Read-only file system") {
+			return ErrTargetIsReadOnly
+		}
+
+		// Other errors, just put them back on stderr
+		_, fpErr := fmt.Fprint(os.Stderr, tarErr.String())
+		if fpErr != nil {
+			return errors.Join(fpErr, err)
+		}
+
+		return errors.Join(fmt.Errorf("failed to wait %v", tarXCmd.Args), err)
 	}
+
 	return nil
 }
 
-func mountSnapshotForContainer(ctx context.Context, client *containerd.Client, container containerd.Container, snapshotter string) (string, func(), error) {
-	cinfo, err := container.Info(ctx)
-	if err != nil {
-		return "", nil, err
-	}
-	snapKey := cinfo.SnapshotKey
+func mountSnapshotForContainer(ctx context.Context, client *containerd.Client, conInfo containers.Container, snapshotter string) (string, func() error, error) {
+	snapKey := conInfo.SnapshotKey
 	resp, err := client.SnapshotService(snapshotter).Mounts(ctx, snapKey)
 	if err != nil {
 		return "", nil, err
 	}
+
```

Comment:
I know you don't like these Akihiro, but would you reconsider? It does make code much easier to read for these old eyes here :-)